src/crimson/osd/pg.cc

```diff
@@ -370,8 +381,44 @@ void PG::prepare_write(pg_info_t &info,
 std::pair<ghobject_t, bool>
 PG::do_delete_work(ceph::os::Transaction &t, ghobject_t _next)
 {
-  // TODO
-  return {_next, false};
+  logger().info("removing pg {}", pgid);
+  auto fut = interruptor::make_interruptible(
+    shard_services.get_store().list_objects(
+      coll_ref,
+      _next,
+      ghobject_t::get_max(),
+      local_conf()->osd_target_transaction_size));
+
+  auto [objs_to_rm, next] = fut.get();
+  if (objs_to_rm.empty()) {
+    logger().info("all objs removed, removing coll for {}", pgid);
+    t.remove(coll_ref->get_cid(), pgmeta_oid);
+    t.remove_collection(coll_ref->get_cid());
+    (void) shard_services.get_store().do_transaction(
```

Comment:
Can we really ignore this future?

Reply:
Honestly speaking, I'm also not quite sure. The classic osd ignores the finish of the collection remove transaction, so I did the same. What do you think? @athanatos

Reply:
@cyx1231st All other comments are addressed, please take a look again.

Reply:
It's fine for now.